lib/AST/SubstitutionMap.cpp

```diff
@@ -64,173 +72,115 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
-          llvm::SmallPtrSetImpl<ProtocolDecl *> &visited) -> Optional<T> {
-    // Only visit a particular protocol once.
-    auto proto = conformance.getRequirement();
-    if (!visited.insert(proto).second) return None;
-
-    // Check whether this protocol has an associated type with the
-    // same name as the one we're looking for.
-    AssociatedTypeDecl *protoAssocType = nullptr;
-    for (auto member : proto->lookupDirect(associatedTypeName)) {
-      protoAssocType = dyn_cast<AssociatedTypeDecl>(member);
-      if (protoAssocType) break;
-    }
+  auto genericSig = getGenericSignature();
+  auto &mod = *proto->getModuleContext();
 
-    if (protoAssocType) {
-      if (conformance.isAbstract()) {
-        // Narrow fix since this whole code path is going away soon
-        // (if you find this code still here a year from now, I will be
-        // very ashamed).
-        //
-        // FIXME: Should always have a valid generic environment.
-        if (protoAssocType->getProtocol()->getGenericEnvironment()) {
-          for (auto assocProto : protoAssocType->getConformingProtocols()) {
-            if (auto found = fn(ProtocolConformanceRef(assocProto)))
-              return found;
-          }
-        }
-      } else {
-        auto sub = conformance.getConcrete()->getTypeWitnessSubstAndDecl(
-                                           protoAssocType, nullptr).first;
-        for (auto subConformance : sub.getConformances()) {
-          if (auto found = fn(subConformance))
-            return found;
-        }
-      }
-    }
+  // HACK: Deal with AnyObject conformances, which get magically dropped in
+  // frustrating ways.
+  // FIXME: This hack dies with AnyObject-as-a-protocol.
+  if (proto->isSpecificProtocol(KnownProtocolKind::AnyObject) &&
+      genericSig->requiresClass(type, mod))
+    return ProtocolConformanceRef(proto);
 
-    // Search inherited conformances.
-    for (auto inherited : proto->getInheritedProtocols()) {
-      if (auto found = searchInConformance(conformance.getInherited(inherited),
-                                           associatedTypeName,
-                                           visited))
-        return found;
-    }
+  // If the type doesn't conform to this protocol, fail.
+  if (!genericSig->conformsToProtocol(type, proto, mod))
     return None;
-  };
 
-  // Check if we have conformances from one of our parent types.
-  return forEachParent<ProtocolConformanceRef>(type, visitedParents,
-      [&](CanType parent, AssociatedTypeDecl *assocType)
-         -> Optional<ProtocolConformanceRef> {
-    return forEachConformance<T>(parent, visitedParents,
-        [&](ProtocolConformanceRef conformance) -> Optional<T> {
-      llvm::SmallPtrSet<ProtocolDecl *, 4> visited;
-      return searchInConformance(conformance, assocType->getFullName(),
-                                 visited);
-    });
-  });
-}
+  auto canonType = genericSig->getCanonicalTypeInContext(type, mod);
+  auto accessPath =
```

Comment:
Actually, `getConformanceAccessPath()` can work on non-canonical types, because it implicitly canonicalizes when looking up the potential archetype. I think the canonicalization you refer to is unnecessary.